File: android/app/src/main/java/io/github/manhvu1212/tallyo/ui/addround/AddRoundScreen.kt
package io.github.manhvu1212.tallyo.ui.addround

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.imePadding
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.foundation.text.KeyboardActions
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.derivedStateOf
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateMapOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.runtime.snapshots.SnapshotStateMap
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.focus.FocusRequester
import androidx.compose.ui.focus.focusRequester
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.ImeAction
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import androidx.lifecycle.viewmodel.compose.viewModel
import io.github.manhvu1212.tallyo.R
import io.github.manhvu1212.tallyo.domain.Player
import io.github.manhvu1212.tallyo.domain.Round
import io.github.manhvu1212.tallyo.domain.RoundScore
import io.github.manhvu1212.tallyo.domain.Session
import io.github.manhvu1212.tallyo.ui.LocalAppContainer
import io.github.manhvu1212.tallyo.ui.components.PrimaryButton
import io.github.manhvu1212.tallyo.ui.components.TallyoTextField
import io.github.manhvu1212.tallyo.ui.components.TopBar
import io.github.manhvu1212.tallyo.ui.theme.TallyoColors

private data class DeltaBtn(val label: String, val value: Int?, val edit: Boolean = false)

private val DELTAS = listOf(
    DeltaBtn("-4", -4),
    DeltaBtn("-2", -2),
    DeltaBtn("-1", -1),
    DeltaBtn("✎", null, edit = true),
    DeltaBtn("+1", 1),
    DeltaBtn("+2", 2),
    DeltaBtn("+4", 4),
)

@Composable
fun AddRoundScreen(
    sessionId: String,
    roundId: String?,
    onBack: () -> Unit,
) {
    val container = LocalAppContainer.current
    val vm: AddRoundViewModel = viewModel(
        factory = AddRoundViewModel.factory(container.repository, sessionId, roundId),
    )
    val session by vm.session.collectAsStateWithLifecycle()

    val current = session
    if (current == null) {
        Column(Modifier.fillMaxSize().background(TallyoColors.Bg)) {
            TopBar(title = stringResource(R.string.round_title_new), onBack = onBack)
            Text(
                stringResource(R.string.round_not_found),
                color = TallyoColors.TextMuted,
                modifier = Modifier.padding(16.dp),
            )
        }
        return
    }

    val editing: Round? = remember(current, roundId) {
        roundId?.let { id -> current.rounds.firstOrNull { it.id == id } }
    }

    val visiblePlayers: List<Player> = remember(current, editing) {
        if (editing != null) {
            val ids = editing.scores.map { it.playerId }.toSet()
            current.players.filter { it.id in ids }
        } else {
            current.players.filter { !it.resting }
        }
    }

    val scores: SnapshotStateMap<String, String> = remember(current, editing) {
        mutableStateMapOf<String, String>().apply {
            visiblePlayers.forEach { p ->
                val s = editing?.scores?.firstOrNull { it.playerId == p.id }
                put(p.id, s?.points?.toString() ?: "")
            }
        }
    }

    var note by remember { mutableStateOf(editing?.note.orEmpty()) }
    var activeId by remember { mutableStateOf<String?>(null) }
    var customMode by remember { mutableStateOf(false) }
    var invalid by remember { mutableStateOf(false) }
    var emptyAll by remember { mutableStateOf(false) }
    var needMore by remember { mutableStateOf(false) }
    var unbalancedBy by remember { mutableStateOf<Int?>(null) }
    val customFocus = remember { FocusRequester() }

    LaunchedEffect(activeId) { customMode = false }

    val liveStatus by remember(scores, visiblePlayers) {
        derivedStateOf {
            var sum = 0
            var emptyId: String? = null
            var emptyCount = 0
            var err = false
            for (p in visiblePlayers) {
                val raw = (scores[p.id] ?: "").trim()
                if (raw.isEmpty()) {
                    if (emptyId == null) emptyId = p.id
                    emptyCount++
                    continue
                }
                val n = raw.toIntOrNull()
                if (n == null) {
                    err = true
                    continue
                }
                sum += n
            }
            LiveStatus(sum, emptyId, emptyCount, err)
        }
    }

    val zeroSum = current.zeroSum

    val onSave: () -> Unit = {
        val ls = liveStatus
        if (ls.parseError) {
            invalid = true
        } else {
            val filledCount = visiblePlayers.size - ls.emptyCount
            if (filledCount == 0) {
                emptyAll = true
            } else if (zeroSum && ls.emptyCount >= 2) {
                needMore = true
            } else if (zeroSum && ls.emptyCount == 0 && ls.sum != 0) {
                unbalancedBy = ls.sum
            } else {
                val built: List<RoundScore> = if (zeroSum) {
                    visiblePlayers.map { p ->
                        val raw = (scores[p.id] ?: "").trim()
                        if (raw.isEmpty()) RoundScore(p.id, -ls.sum)
                        else RoundScore(p.id, raw.toInt())
                    }
                } else {
                    visiblePlayers.map { p ->
                        val raw = (scores[p.id] ?: "").trim()
                        RoundScore(p.id, if (raw.isEmpty()) 0 else raw.toInt())
                    }
                }
                val noteFinal = note.trim().takeIf { it.isNotEmpty() }
                if (editing == null) vm.saveNew(built, noteFinal, onBack)
                else vm.saveEdit(built, noteFinal, onBack)
            }
        }
    }

    Column(Modifier.fillMaxSize().background(TallyoColors.Bg)) {
        TopBar(
            title = if (editing != null) stringResource(R.string.round_title_edit)
            else stringResource(R.string.round_title_new),
            onBack = onBack,
        )
        Column(
            modifier = Modifier
                .fillMaxSize()
                .imePadding()
                .verticalScroll(rememberScrollState())
                .padding(16.dp),
        ) {
            Text(
                if (zeroSum) stringResource(R.string.round_hint_zero_sum)
                else stringResource(R.string.round_hint_free),
                color = TallyoColors.TextMuted,
                fontSize = 13.sp,
                lineHeight = 20.sp,
            )

            Spacer(Modifier.height(12.dp))
            Column(verticalArrangement = Arrangement.spacedBy(12.dp)) {
                visiblePlayers.forEach { p ->
                    val raw = scores[p.id] ?: ""
                    val isActive = activeId == p.id
                    val isCustom = isActive && customMode
                    val isAutoFilled = zeroSum && liveStatus.emptyCount == 1 && liveStatus.emptyId == p.id
                    val autoVal = -liveStatus.sum

                    val rawNum = raw.toIntOrNull()
                    val valueColor = when {
                        raw.isEmpty() && isAutoFilled -> TallyoColors.Primary
                        raw.isEmpty() -> TallyoColors.TextMuted
                        rawNum != null && rawNum > 0 -> TallyoColors.Win
                        rawNum != null && rawNum < 0 -> TallyoColors.Loss
                        else -> TallyoColors.Text
                    }

                    val displayValue = when {
                        raw.isNotEmpty() -> raw
                        isAutoFilled -> "= ${if (autoVal > 0) "+$autoVal" else "$autoVal"}"
                        else -> "0"
                    }

                    Column {
                        val rowShape = RoundedCornerShape(10.dp)
                        Row(
                            modifier = Modifier
                                .fillMaxWidth()
                                .clip(rowShape)
                                .background(if (isActive) TallyoColors.PrimaryTintBg else TallyoColors.Surface)
                                .border(1.dp, if (isActive) TallyoColors.Primary else TallyoColors.Border, rowShape)
                                .clickable { activeId = p.id }
                                .padding(horizontal = 12.dp, vertical = 8.dp),
                            verticalAlignment = Alignment.CenterVertically,
                        ) {
                            Text(p.name, color = TallyoColors.Text, fontSize = 16.sp, fontWeight = FontWeight.Medium, modifier = Modifier.weight(1f))
                            if (isCustom) {
                                BasicTextField(
                                    value = raw,
                                    onValueChange = { scores[p.id] = it },
                                    singleLine = true,
                                    textStyle = TextStyle(
                                        color = valueColor,
                                        fontSize = 18.sp,
                                        fontWeight = FontWeight.Bold,
                                        textAlign = TextAlign.End,
                                    ),
                                    keyboardOptions = KeyboardOptions(
                                        keyboardType = KeyboardType.NumberPassword,
                                        imeAction = ImeAction.Done,
                                    ),
                                    keyboardActions = KeyboardActions(onDone = { customMode = false }),
                                    cursorBrush = SolidColor(TallyoColors.Primary),
                                    modifier = Modifier
                                        .focusRequester(customFocus)
                                        .padding(vertical = 4.dp),
                                )
                                LaunchedEffect(customMode, activeId) {
                                    if (isCustom) customFocus.requestFocus()
                                }
                            } else {
                                Text(
                                    displayValue,
                                    color = valueColor,
                                    fontSize = 18.sp,
                                    fontWeight = FontWeight.Bold,
                                    textAlign = TextAlign.End,
                                )
                            }
                        }

                        if (isActive && !isCustom) {
                            Spacer(Modifier.height(4.dp))
                            Row(
                                horizontalArrangement = Arrangement.spacedBy(4.dp),
                                modifier = Modifier.fillMaxWidth(),
                            ) {
                                DELTAS.forEach { d ->
                                    val isEdit = d.edit
                                    val isNeg = d.value != null && d.value < 0
                                    val isPos = d.value != null && d.value > 0
                                    val btnShape = RoundedCornerShape(10.dp)
                                    Box(
                                        modifier = Modifier
                                            .weight(1f)
                                            .clip(btnShape)
                                            .background(if (isEdit) TallyoColors.Primary else TallyoColors.SurfaceAlt)
                                            .border(1.dp, if (isEdit) TallyoColors.Primary else TallyoColors.Border, btnShape)
                                            .clickable {
                                                if (isEdit) customMode = true
                                                else d.value?.let { delta ->
                                                    val cur = (scores[p.id] ?: "").trim().toIntOrNull() ?: 0
                                                    scores[p.id] = (cur + delta).toString()
                                                }
                                            }
                                            .padding(vertical = 12.dp),
                                        contentAlignment = Alignment.Center,
                                    ) {
                                        Text(
                                            d.label,
                                            color = when {
                                                isEdit -> Color.White
                                                isNeg -> TallyoColors.Loss
                                                isPos -> TallyoColors.Win
                                                else -> TallyoColors.Text
                                            },
                                            fontSize = 16.sp,
                                            fontWeight = FontWeight.Bold,
                                        )
                                    }
                                }
                            }
                        }
                    }
                }
            }

            Spacer(Modifier.height(24.dp))
            Text(stringResource(R.string.round_note_label), color = TallyoColors.TextMuted, fontSize = 13.sp)
            Spacer(Modifier.height(8.dp))
            TallyoTextField(
                value = note,
                onValueChange = { note = it },
                placeholder = stringResource(R.string.round_note_placeholder),
                modifier = Modifier.fillMaxWidth(),
            )

            Spacer(Modifier.height(24.dp))
            PrimaryButton(
                label = if (editing != null) stringResource(R.string.round_save_edit)
                else stringResource(R.string.round_save_new),
                onClick = onSave,
                modifier = Modifier.fillMaxWidth(),
            )
            Spacer(Modifier.height(48.dp))
        }
    }

    SimpleAlert(
        open = invalid,
        title = R.string.round_invalid_title,
        message = R.string.round_invalid_message,
        onDismiss = { invalid = false },
    )
    SimpleAlert(
        open = emptyAll,
        title = R.string.round_empty_title,
        message = R.string.round_empty_message,
        onDismiss = { emptyAll = false },
    )
    if (needMore) {
        AlertDialog(
            onDismissRequest = { needMore = false },
            title = { Text(stringResource(R.string.round_need_more_title), color = TallyoColors.Text) },
            text = {
                Text(
                    stringResource(R.string.round_need_more_message, visiblePlayers.size - 1),
                    color = TallyoColors.TextMuted,
                )
            },
            containerColor = TallyoColors.Surface,
            confirmButton = {
                TextButton(onClick = { needMore = false }) {
                    Text("OK", color = TallyoColors.Primary)
                }
            },
        )
    }
    unbalancedBy?.let { sum ->
        val display = if (sum > 0) "+$sum" else "$sum"
        AlertDialog(
            onDismissRequest = { unbalancedBy = null },
            title = { Text(stringResource(R.string.round_unbalanced_title), color = TallyoColors.Text) },
            text = {
                Text(stringResource(R.string.round_unbalanced_message, display), color = TallyoColors.TextMuted)
            },
            containerColor = TallyoColors.Surface,
            confirmButton = {
                TextButton(onClick = { unbalancedBy = null }) {
                    Text("OK", color = TallyoColors.Primary)
                }
            },
        )
    }
}

private data class LiveStatus(
    val sum: Int,
    val emptyId: String?,
    val emptyCount: Int,
    val parseError: Boolean,
)

@Composable
private fun SimpleAlert(
    open: Boolean,
    title: Int,
    message: Int,
    onDismiss: () -> Unit,
) {
    if (!open) return
    AlertDialog(
        onDismissRequest = onDismiss,
        title = { Text(stringResource(title), color = TallyoColors.Text) },
        text = { Text(stringResource(message), color = TallyoColors.TextMuted) },
        containerColor = TallyoColors.Surface,
        confirmButton = {
            TextButton(onClick = onDismiss) {
                Text("OK", color = TallyoColors.Primary)
            }
        },
    )
}

File: ios/Tallyo/UI/Screens/AddRoundScreen.swift
import SwiftUI
import SwiftData

private struct DeltaBtn: Identifiable {
    let id = UUID()
    let label: String
    let value: Int?
    let edit: Bool
}

private let DELTAS = [
    DeltaBtn(label: "-4", value: -4, edit: false),
    DeltaBtn(label: "-2", value: -2, edit: false),
    DeltaBtn(label: "-1", value: -1, edit: false),
    DeltaBtn(label: "✎", value: nil, edit: true),
    DeltaBtn(label: "+1", value: 1, edit: false),
    DeltaBtn(label: "+2", value: 2, edit: false),
    DeltaBtn(label: "+4", value: 4, edit: false)
]

struct AddRoundScreen: View {
    @Environment(\.modelContext) private var modelContext
    @Environment(\.colorScheme) private var colorScheme
    
    let sessionId: String
    let roundId: String?
    let onBack: () -> Void
    
    @State private var scores: [String: String] = [:] // playerId -> text
    @State private var note: String = ""
    @State private var activeId: String? = nil
    @State private var customMode: Bool = false
    
    @State private var showInvalidAlert = false
    @State private var showEmptyAlert = false
    @State private var showNeedMoreAlert = false
    @State private var showUnbalancedAlert = false
    @State private var unbalancedSum = 0
    
    var body: some View {
        // Query the session details
        let descriptor = FetchDescriptor<Session>(
            predicate: #Predicate<Session> { $0.id == sessionId }
        )
        let sessions = try? modelContext.fetch(descriptor)
        guard let session = sessions?.first else {
            return AnyView(
                VStack {
                    Text(NSLocalizedString("round_not_found", comment: ""))
                    Button("Back", action: onBack)
                }
            )
        }
        
        let editing = roundId.flatMap { rid in session.rounds.first { $0.id == rid } }
        
        let visiblePlayers = editing != nil ?
            session.players.filter { p in editing!.scores.contains { $0.playerId == p.id } } :
            session.players.filter { !$0.resting }
            
        let pendingEvents = session.events.filter { $0.roundId == nil }
        let zeroSum = session.zeroSum
        let palette = TallyoTheme.palette(for: colorScheme)
        
        // Compute live values
        var sum = 0
        var emptyId: String? = nil
        var emptyCount = 0
        var parseError = false
        
        for p in visiblePlayers {
            let pendingPoints = editing == nil ? pendingEvents.filter { $0.playerId == p.id }.reduce(0) { $0 + $1.points } : 0
            let raw = (scores[p.id] ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
            if raw.isEmpty {
                if emptyId == nil { emptyId = p.id }
                emptyCount += 1
                sum += pendingPoints
                continue
            }
            if let n = Int(raw) {
                sum += (n + pendingPoints)
            } else {
                parseError = true
            }
        }
        
        let liveSum = sum
        let liveEmptyId = emptyId
        let liveEmptyCount = emptyCount
        let liveParseError = parseError
        
        let unbalanced = zeroSum && liveEmptyCount == 0 && liveSum != 0 && !liveParseError
        
        let mainContent = NavigationStack {
            ZStack {
                palette.bg.ignoresSafeArea()
                
                VStack(spacing: 0) {
                    // Top navigation bar
                    HStack {
                        Button(action: onBack) {
                            Image(systemName: "chevron.left")
                                .font(.system(size: 20, weight: .semibold))
                                .foregroundColor(palette.text)
                        }
                        
                        Spacer()
                        
                        Text(editing != nil ? NSLocalizedString("round_title_edit", comment: "") : NSLocalizedString("round_title_new", comment: ""))
                            .font(.system(size: 18, weight: .bold))
                            .foregroundColor(palette.text)
                        
                        Spacer()
                        
                        Color.clear.frame(width: 24, height: 24)
                    }
                    .padding(.horizontal, 16)
                    .padding(.vertical, 12)
                    
                    ScrollView {
                        VStack(alignment: .leading, spacing: 20) {
                            
                            // Hint
                            Text(zeroSum ? NSLocalizedString("round_hint_zero_sum", comment: "") : NSLocalizedString("round_hint_free", comment: ""))
                                .font(.system(size: 13))
                                .foregroundColor(palette.textMuted)
                                .lineSpacing(4)
                                .padding(.horizontal, 4)
                            
                            // Pending Quick Events Banner
                            if editing == nil && !pendingEvents.isEmpty {
                                let pendingSummary = buildPendingSummary(pendingEvents: pendingEvents, players: session.players)
                                VStack(alignment: .leading, spacing: 4) {
                                    Text(String(format: NSLocalizedString("add_round_pending_banner", comment: ""), pendingEvents.count, pendingSummary))
                                        .font(.system(size: 13, weight: .medium))
                                        .foregroundColor(palette.primary)
                                        .lineSpacing(3)
                                }
                                .padding(12)
                                .background(palette.primaryTintBg)
                                .cornerRadius(8)
                                .overlay(
                                    RoundedRectangle(cornerRadius: 8)
                                        .stroke(palette.primary.opacity(0.5), lineWidth: 1)
                                )
                            }
                            
                            // Player Input Cards
                            VStack(spacing: 12) {
                                ForEach(visiblePlayers) { p in
                                    let raw = scores[p.id] ?? ""
                                    let isActive = activeId == p.id
                                    let isCustom = isActive && customMode
                                    let isAutoFilled = zeroSum && liveEmptyCount == 1 && liveEmptyId == p.id
                                    let autoVal = -liveSum
                                    let showUnbalanced = unbalanced && isActive
                                    
                                    let rawNum = Int(raw)
                                    
                                    let valueColor: Color = {
                                        if raw.isEmpty && isAutoFilled { return palette.primary }
                                        if raw.isEmpty { return palette.textMuted }
                                        if let num = rawNum, num > 0 { return palette.win }
                                        if let num = rawNum, num < 0 { return palette.loss }
                                        return palette.text
                                    }()
                                    
                                    let displayValue: String = {
                                        if !raw.isEmpty { return raw }
                                        if isAutoFilled { return "= \(autoVal > 0 ? "+" : "")\(autoVal)" }
                                        return "0"
                                    }()
                                    
                                    let pendingPoints = editing == nil ? pendingEvents.filter { $0.playerId == p.id }.reduce(0) { $0 + $1.points } : 0
                                    let baseScore = !raw.isEmpty ? (Int(raw) ?? 0) : (isAutoFilled ? autoVal : 0)
                                    let totalScore = baseScore + pendingPoints
                                    
                                    VStack(spacing: 4) {
                                        Button(action: {
                                            activeId = p.id
                                            customMode = false
                                        }) {
                                            HStack {
                                                VStack(alignment: .leading, spacing: 4) {
                                                    HStack {
                                                        Text(p.name)
                                                            .font(.system(size: 16, weight: .medium))
                                                            .foregroundColor(palette.text)
                                                        
                                                        if editing == nil && pendingPoints != 0 {
                                                            Text("⚡ \(pendingPoints > 0 ? "+" : "")\(pendingPoints)")
                                                                .font(.system(size: 11, weight: .bold))
                                                                .padding(.horizontal, 6)
                                                                .padding(.vertical, 2)
                                                                .background((pendingPoints > 0 ? palette.win : palette.loss).opacity(0.15))
                                                                .foregroundColor(pendingPoints > 0 ? palette.win : palette.loss)
                                                                .cornerRadius(4)
                                                        }
                                                    }
                                                }
                                                
                                                Spacer()
                                                
                                                VStack(alignment: .trailing, spacing: 2) {
                                                    if isCustom {
                                                        TextField("", text: Binding(
                                                            get: { scores[p.id] ?? "" },
                                                            set: { scores[p.id] = $0 }
                                                        ))
                                                        .keyboardType(.numbersAndPunctuation)
                                                        .multilineTextAlignment(.trailing)
                                                        .font(.system(size: 18, weight: .bold))
                                                        .foregroundColor(valueColor)
                                                        .frame(width: 80)
                                                    } else {
                                                        Text(displayValue)
                                                            .font(.system(size: 18, weight: .bold))
                                                            .foregroundColor(valueColor)
                                                    }
                                                    
                                                    if editing == nil && pendingPoints != 0 {
                                                        Text("\(NSLocalizedString("stats_table_total", comment: "")): \(totalScore > 0 ? "+" : "")\(totalScore)")
                                                            .font(.system(size: 12))
                                                            .foregroundColor(palette.textMuted)
                                                    }
                                                }
                                            }
                                            .padding(.horizontal, 12)
                                            .padding(.vertical, 10)
                                            .background(isActive ? palette.primaryTintBg : palette.surface)
                                            .cornerRadius(10)
                                            .overlay(
                                                RoundedRectangle(cornerRadius: 10)
                                                    .stroke(showUnbalanced ? palette.danger : (isActive ? palette.primary : palette.border), lineWidth: showUnbalanced ? 2 : 1)
                                            )
                                        }
                                        .buttonStyle(.plain)
                                        
                                        // Delta controls if active
                                        if isActive && !isCustom {
                                            HStack(spacing: 4) {
                                                ForEach(DELTAS) { d in
                                                    Button(action: {
                                                        if d.edit {
                                                            customMode = true
                                                        } else if let delta = d.value {
                                                            let cur = Int(scores[p.id] ?? "") ?? 0
                                                            scores[p.id] = "\(cur + delta)"
                                                        }
                                                    }) {
                                                        Text(d.label)
                                                            .font(.system(size: 15, weight: .bold))
                                                            .foregroundColor(d.edit ? .white : (d.value != nil && d.value! < 0 ? palette.loss : (d.value != nil && d.value! > 0 ? palette.win : palette.text)))
                                                            .frame(maxWidth: .infinity)
                                                            .padding(.vertical, 10)
                                                            .background(d.edit ? palette.primary : palette.surfaceAlt)
                                                            .cornerRadius(8)
                                                            .overlay(
                                                                RoundedRectangle(cornerRadius: 8)
                                                                    .stroke(d.edit ? palette.primary : palette.border, lineWidth: 1)
                                                            )
                                                    }
                                                    .buttonStyle(.plain)
                                                }
                                            }
                                            .padding(.top, 2)
                                        }
                                    }
                                }
                            }
                            
                            // Note
                            VStack(alignment: .leading, spacing: 6) {
                                Text(NSLocalizedString("round_note_label", comment: ""))
                                    .font(.system(size: 13))
                                    .foregroundColor(palette.textMuted)
                                
                                TextField(NSLocalizedString("round_note_placeholder", comment: ""), text: $note)
                                    .padding(12)
                                    .background(palette.surface)
                                    .foregroundColor(palette.text)
                                    .cornerRadius(8)
                                    .overlay(
                                        RoundedRectangle(cornerRadius: 8)
                                            .stroke(palette.border, lineWidth: 1)
                                    )
                                    .textInputAutocapitalization(.sentences)
                            }
                            
                            // Save Button
                            Button(action: {
                                saveRound(session: session, visiblePlayers: visiblePlayers, liveSum: liveSum, liveEmptyId: liveEmptyId, liveEmptyCount: liveEmptyCount, liveParseError: liveParseError)
                            }) {
                                Text(editing != nil ? NSLocalizedString("round_save_edit", comment: "") : NSLocalizedString("round_save_new", comment: ""))
                                    .font(.system(size: 16, weight: .bold))
                                    .frame(maxWidth: .infinity)
                                    .padding(.vertical, 14)
                                    .background(palette.primary)
                                    .foregroundColor(.white)
                                    .cornerRadius(8)
                            }
                            .padding(.top, 16)
                            .padding(.bottom, 48)
                        }
                        .padding(.horizontal, 16)
                    }
                }
            }
            .navigationBarHidden(true)
            .alert(NSLocalizedString("round_invalid_title", comment: ""), isPresented: $showInvalidAlert) {
                Button("OK", role: .cancel) { }
            } message: {
                Text(NSLocalizedString("round_invalid_message", comment: ""))
            }
            .alert(NSLocalizedString("round_empty_title", comment: ""), isPresented: $showEmptyAlert) {
                Button("OK", role: .cancel) { }
            } message: {
                Text(NSLocalizedString("round_empty_message", comment: ""))
            }
            .alert(NSLocalizedString("round_need_more_title", comment: ""), isPresented: $showNeedMoreAlert) {
                Button("OK", role: .cancel) { }
            } message: {
                Text(String(format: NSLocalizedString("round_need_more_message", comment: ""), visiblePlayers.count - 1))
            }
            .alert(NSLocalizedString("round_unbalanced_title", comment: ""), isPresented: $showUnbalancedAlert) {
                Button("OK", role: .cancel) { }
            } message: {
                let display = unbalancedSum > 0 ? "+\(unbalancedSum)" : "\(unbalancedSum)"
                Text(String(format: NSLocalizedString("round_unbalanced_message", comment: ""), display))
            }
            .onAppear {
                if let editing = editing {
                    for score in editing.scores {
                        scores[score.playerId] = "\(score.points)"
                    }
                    note = editing.note ?? ""
                }
            }
        }
        
        return AnyView(mainContent)
    }
    
    private func buildPendingSummary(pendingEvents: [RoundEvent], players: [Player]) -> String {
        let grouped = groupEvents(events: pendingEvents, players: players)
        return grouped.map { item in
            switch item {
            case .individual(let pe):
                let playerName = players.first { $0.id == pe.playerId }?.name ?? ""
                let pointsFormatted = "\(pe.points > 0 ? "+" : "")\(pe.points)"
                let noteStr = pe.note.flatMap { !$0.isEmpty ? " (\($0))" : "" } ?? ""
                return "\(playerName) \(pointsFormatted)\(noteStr)"
            case .transfer(let fromEvent, let toEvent, let baseNote):
                let fromPlayerName = players.first { $0.id == fromEvent.playerId }?.name ?? ""
                let toPlayerName = players.first { $0.id == toEvent.playerId }?.name ?? ""
                let pointsFormatted = "\(toEvent.points > 0 ? "+" : "")\(toEvent.points)"
                let noteStr = !baseNote.isEmpty ? " (\(baseNote))" : ""
                return "\(fromPlayerName) → \(toPlayerName) \(pointsFormatted)\(noteStr)"
            }
        }.joined(separator: ", ")
    }
    
    private func saveRound(session: Session, visiblePlayers: [Player], liveSum: Int, liveEmptyId: String?, liveEmptyCount: Int, liveParseError: Bool) {
        if liveParseError {
            showInvalidAlert = true
            return
        }
        
        let filledCount = visiblePlayers.count - liveEmptyCount
        if filledCount == 0 {
            showEmptyAlert = true
            return
        }
        
        let zeroSum = session.zeroSum
        
        if zeroSum && liveEmptyCount >= 2 {
            showNeedMoreAlert = true
            return
        }
        
        if zeroSum && liveEmptyCount == 0 && liveSum != 0 {
            unbalancedSum = liveSum
            showUnbalancedAlert = true
            return
        }
        
        // Save
        let finalScores: [Score]
        if zeroSum {
            finalScores = visiblePlayers.map { p in
                let raw = (scores[p.id] ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
                let points = raw.isEmpty ? -liveSum : (Int(raw) ?? 0)
                return Score(roundId: roundId ?? UUID().uuidString, playerId: p.id, points: points)
            }
        } else {
            finalScores = visiblePlayers.map { p in
                let raw = (scores[p.id] ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
                let points = raw.isEmpty ? 0 : (Int(raw) ?? 0)
                return Score(roundId: roundId ?? UUID().uuidString, playerId: p.id, points: points)
            }
        }
        
        let finalNote = note.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ? nil : note.trimmingCharacters(in: .whitespacesAndNewlines)
        
        if let rid = roundId {
            // Update round
            if let round = session.rounds.first(where: { $0.id == rid }) {
                round.note = finalNote
                
                // Clear old scores and add new
                for score in round.scores {
                    modelContext.delete(score)
                }
                round.scores.removeAll()
                
                for fs in finalScores {
                    fs.round = round
                    round.scores.append(fs)
                }
            }
        } else {
            // New round
            let newRound = Round(note: finalNote, orderIndex: session.rounds.count)
            modelContext.insert(newRound)
            newRound.session = session
            session.rounds.append(newRound)
            
            for fs in finalScores {
                fs.round = newRound
                newRound.scores.append(fs)
            }
            
            // Move pending events into the round
            let pendingEvents = session.events.filter { $0.roundId == nil }
            for pe in pendingEvents {
                pe.round = newRound
                newRound.events.append(pe)
            }
        }
        
        session.updatedAt = Date()
        try? modelContext.save()
        onBack()
    }
}
